<a href="https://colab.research.google.com/github/Joez93/Joez93.github.io/blob/master/Titanic_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Write by
# RENPENG ZHOU

In [0]:
import numpy as np
import pandas as pd
import statistics 
from numpy import median
from numpy import mean

Import data to first dataframe

In [0]:
# Import data from Web to create a dateframe
df = pd.read_csv("https://raw.githubusercontent.com/Joez93/Joez93.github.io/594b6de6d7b2f0183acce42e42572267e38c0f2b/train.csv")

Check data type, **Numerical** or **Categorical**

In [161]:
# See all date type of the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [162]:
# Get a basic statistic info of the dataframe
df.describe()

PassengerId    Survived      Pclass  ...       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  ...  891.000000  891.000000  891.000000
mean    446.000000    0.383838    2.308642  ...    0.523008    0.381594   32.204208
std     257.353842    0.486592    0.836071  ...    1.102743    0.806057   49.693429
min       1.000000    0.000000    1.000000  ...    0.000000    0.000000    0.000000
25%     223.500000    0.000000    2.000000  ...    0.000000    0.000000    7.910400
50%     446.000000    0.000000    3.000000  ...    0.000000    0.000000   14.454200
75%     668.500000    1.000000    3.000000  ...    1.000000    0.000000   31.000000
max     891.000000    1.000000    3.000000  ...    8.000000    6.000000  512.329200

[8 rows x 7 columns]

891 people on borad, and from the mean of Survived columns we can know Only 38.3838% people survived

In [163]:
# Find all survivors and count total
survival = df.loc[(df['Survived'] == 1)]
print ('Total people survived is ' + str(len(survival)))
# Double Check surivial rate
survial_rate = float(round(342/891, 3))
print("The survivial rate is " + str(survial_rate*100) + "%")

Total people survived is 342
The survivial rate is 38.4%




---


# Find NaN values from dataframe and replacing them




In [164]:
# Find total number of missing value of each columns
NaN_value = df.isnull().sum().sort_values(ascending = False)
NaN_value

Cabin          687
Age            177
Embarked         2
Fare             0
Ticket           0
Parch            0
SibSp            0
Sex              0
Name             0
Pclass           0
Survived         0
PassengerId      0
dtype: int64



# 1.   Exclude attribute "cabin" since it has too many vacancies to make reasonable analysis
# 2.   Exclude attribute "Ticket" since the ticket numbers were just random reference numbers, and they are meaningless to survival rate





In [165]:
# Delete column "cabin" since it has too many vacancies
# Delete column "Ticket" since the ticket number is meaningless to survival rate
df_2 = df.drop(columns = ['Cabin','Ticket'])
# Testing
df_2.head(3)

PassengerId  Survived  Pclass  ... Parch     Fare  Embarked
0            1         0       3  ...     0   7.2500         S
1            2         1       1  ...     0  71.2833         C
2            3         1       3  ...     0   7.9250         S

[3 rows x 10 columns]

# Modify dataframe

Replace all vacancies of age with the average age of remaining

In [166]:
# Modifing "Age" column
Ages = df_2.loc[:,'Age']
Average_Age = round(Ages.mean(),0)  # Find average age
print("The average age is " + str(Average_Age))
Ages.fillna(Average_Age, inplace = True) 
df_2.replace(df_2['Age'], pd.Series(Ages))

The average age is 30.0


PassengerId  Survived  Pclass  ... Parch     Fare  Embarked
0              1         0       3  ...     0   7.2500         S
1              2         1       1  ...     0  71.2833         C
2              3         1       3  ...     0   7.9250         S
3              4         1       1  ...     0  53.1000         S
4              5         0       3  ...     0   8.0500         S
..           ...       ...     ...  ...   ...      ...       ...
886          887         0       2  ...     0  13.0000         S
887          888         1       1  ...     0  30.0000         S
888          889         0       3  ...     2  23.4500         S
889          890         1       1  ...     0  30.0000         C
890          891         0       3  ...     0   7.7500         Q

[891 rows x 10 columns]

In [167]:
# Change data type of 'Age' to integer
df_2['Age'] = df_2['Age'].astype(np.int64)
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            891 non-null int64
SibSp          891 non-null int64
Parch          891 non-null int64
Fare           891 non-null float64
Embarked       889 non-null object
dtypes: float64(1), int64(6), object(3)
memory usage: 69.7+ KB


In [168]:
print("Counts each Embarked")
df_2['Embarked'].value_counts()

Counts each Embarked


S    644
C    168
Q     77
Name: Embarked, dtype: int64

Replace the 2 mississ entries of 'Embarked' with major Embark 'S' beacasue 'S' is dorminte in numbers(644 out of 891)

In [169]:
Major_Embark = 'S'
Embarked = df_2.loc[:,'Embarked']
Embarked.fillna(Major_Embark, inplace = True)
df_2.replace(df_2['Embarked'],pd.Series(Embarked))
df_2['Embarked'].value_counts()

S    646
C    168
Q     77
Name: Embarked, dtype: int64



---


# Analyzing features of Survivial rate 

#1. **Gender**.

In [170]:
# The correlation between gendar and surivial rate
Table_1 = df_2[['Sex', 'Survived']].groupby(['Sex'], as_index=True).mean().sort_values(by='Survived', ascending=False)
Table_1

Survived
Sex             
female  0.742038
male    0.188908

Observation: 
*   # Female has much higher survival rate : 74.20%
*   Male has much lower survival rate : 18.89%



---



# 2. **Age**.

In [0]:
#Create a new attribute column defined by different age stages of passenger
df_2['Age Groups'] = df_2['Age']   # New column named 
df_2 = df_2[['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 
             'Age Groups', 'SibSp','Parch', 'Fare', 'Embarked']]  # Set new column order
df_2.loc[df_2['Age'] <= 14, 'Age Groups'] = 'Children (0-14)'      
df_2.loc[(df_2['Age'] > 14) & (df_2['Age'] <= 24), 'Age Groups'] = 'Teenager (15-24)'
df_2.loc[(df_2['Age'] > 24) & (df_2['Age'] <= 50), 'Age Groups'] = 'Adult (25-50)'
df_2.loc[(df_2['Age'] > 50) & (df_2['Age'] <= 100), 'Age Groups'] = 'Elder (51-100)'

In [172]:
df_2['Age Groups'].value_counts()
# Age structure

Adult (25-50)       549
Teenager (15-24)    200
Children (0-14)      78
Elder (51-100)       64
Name: Age Groups, dtype: int64

In [173]:
# The correlation between age and surivial rate
Table_2 = df_2[['Age Groups','Survived']].groupby(['Age Groups'],as_index=True).mean().sort_values(by='Survived', ascending=False)
Table_2

Survived
Age Groups                
Children (0-14)   0.576923
Adult (25-50)     0.367942
Teenager (15-24)  0.365000
Elder (51-100)    0.343750

Observation



# *   Children aged 0 to 14 have the highest survival rates : 57.7%
*  Adult aged 25 to 50 have second highest survival rate : 36.8%
*   Teengager aged 15 to 24 have second lowest survial rate : 36.5% 
*   Elder aged more than 51(include 51)years old have the lowest survival rate : 34.4%



---





# 3. **Passenger Class and ticket fare.**

In [174]:
#Number of passenger in each class
df_2['Pclass'].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

Number of passenger in Class 3 is more than total number of Class 1 and Class 2 passenger combined

In [175]:
# The correlation between Passenger class level and fare
Table_3a = df_2[['Pclass','Fare']].groupby('Pclass',as_index=True).mean().sort_values(by='Fare',ascending=False)
Table_3a

Fare
Pclass           
1       84.154687
2       20.662183
3       13.675550

Observed a Very strong positive correlation between Class and fare which make sense since better class level cost more

In [176]:
# The correlation between Passenger class level and surivial rate
Table_3b = df_2[['Pclass','Survived']].groupby('Pclass', as_index=True).mean().sort_values(by='Survived', ascending=False)
Table_3b

Survived
Pclass          
1       0.629630
2       0.472826
3       0.242363

In [188]:
# Compare average fare between survived and unsurvived passenger
Table_3c = df_2[['Fare','Survived']].groupby('Survived', as_index=True).mean().sort_values(by='Fare', ascending=False)
Table_3c

Fare
Survived           
1         48.395408
0         22.117887

Observation: 
# Higher Class level (or Fare), higher survival rate.


---



# 4. **Number Sibling and Spouse on board.**

In [178]:
# The correlation between number of Siblings or spouse and surivial rate
Table_4 = df_2[['SibSp','Survived']].groupby('SibSp', as_index=True).mean().sort_values(by='Survived',ascending=False)
Table_4

Survived
SibSp          
1      0.535885
2      0.464286
0      0.345395
3      0.250000
4      0.166667
5      0.000000
8      0.000000


# * No strong correlation observed


---



# 5. **Number of Parents and children on board.**

In [179]:
#The correlation between Parch number and surivial rate
Table_5 = df_2[['Parch','Survived']].groupby('Parch',as_index=True).mean().sort_values(by='Survived',ascending=False)
Table_5

Survived
Parch          
3      0.600000
1      0.550847
2      0.500000
0      0.343658
5      0.200000
4      0.000000
6      0.000000


* No strong correlation observed


---

# 6. Port of Embarkation vs surivial rate

In [190]:
#The correlation between Port of Embarkation and surivial rate
Table_6 = df_2[['Embarked','Survived']].groupby('Embarked',as_index=True).mean().sort_values(by='Survived',ascending=False)
Table_6

Survived
Embarked          
C         0.553571
Q         0.389610
S         0.339009

# Analyzing survived rate considering **TWO OR MORE** factors 

# 7. Age and Gender



In [191]:
#The surivial rate related to Gener and Age
Table_7 = df_2[['Age Groups','Sex','Survived']].groupby(['Sex','Age Groups'],
                                                        as_index=True).mean().sort_values(by='Survived',ascending=False)
Table_7

Survived
Sex    Age Groups                
female Elder (51-100)    0.941176
       Teenager (15-24)  0.782051
       Adult (25-50)     0.733333
       Children (0-14)   0.615385
male   Children (0-14)   0.538462
       Adult (25-50)     0.189702
       Elder (51-100)    0.127660
       Teenager (15-24)  0.098361

Female passengers had an **absolutely** high survival rate
Child passengers have a **relatively** high survival rate



---


# 8.  Gender and P-Class

In [192]:
#The surivial rate related to Gener and passenger class level
Table_8 = df_2[['Sex','Pclass','Survived']].groupby(['Sex','Pclass'],
                                          as_index=True).mean().sort_values(by='Survived',ascending=False)
Table_8

Survived
Sex    Pclass          
female 1       0.968085
       2       0.921053
       3       0.500000
male   1       0.368852
       2       0.157407
       3       0.135447

Female passengers had an **absolutely** high survival rate
Passengers in Higher class have a **relatively** higher survival rate



---


#9.  Passenger class and Age groups

In [193]:
#The surivial rate related to Age and Passenger class level
Table_9a = df_2[['Age Groups','Pclass','Survived']].groupby(['Age Groups','Pclass'],
                                                            as_index=False).mean().sort_values(by='Survived',ascending=False)
Table_9a

Age Groups  Pclass  Survived
4    Children (0-14)       2  1.000000
3    Children (0-14)       1  0.800000
9   Teenager (15-24)       1  0.794118
0      Adult (25-50)       1  0.630435
6     Elder (51-100)       1  0.461538
10  Teenager (15-24)       2  0.435897
1      Adult (25-50)       2  0.432432
5    Children (0-14)       3  0.407407
11  Teenager (15-24)       3  0.228346
2      Adult (25-50)       3  0.223333
7     Elder (51-100)       2  0.200000
8     Elder (51-100)       3  0.100000

Noticed that some data were found abnormal, like 'Children in class 2 have 100% survival rate'.

In [194]:
# Check the sample size of each dataset to make sure the survival rate is reasonable.
Table_9b = df_2[['Age Groups','Pclass','Survived']].groupby(['Age Groups','Pclass'],
                                                             as_index=False).sum().sort_values(by='Survived',ascending=False)
print(Table_9b)
Table_9c = df_2[['Age Groups','Pclass','Survived']].groupby(['Age Groups','Pclass'],
                                                        as_index=False).count().sort_values(by='Survived',ascending=False)
print(Table_9c)

          Age Groups  Pclass  Survived
0      Adult (25-50)       1        87
2      Adult (25-50)       3        67
1      Adult (25-50)       2        48
11  Teenager (15-24)       3        29
9   Teenager (15-24)       1        27
5    Children (0-14)       3        22
4    Children (0-14)       2        19
6     Elder (51-100)       1        18
10  Teenager (15-24)       2        17
3    Children (0-14)       1         4
7     Elder (51-100)       2         3
8     Elder (51-100)       3         1
          Age Groups  Pclass  Survived
2      Adult (25-50)       3       300
0      Adult (25-50)       1       138
11  Teenager (15-24)       3       127
1      Adult (25-50)       2       111
5    Children (0-14)       3        54
6     Elder (51-100)       1        39
10  Teenager (15-24)       2        39
9   Teenager (15-24)       1        34
4    Children (0-14)       2        19
7     Elder (51-100)       2        15
8     Elder (51-100)       3        10
3    Children (0-14)     

The sample size of few groups are too small, for example: (Elder in Class 2 and 3; Children in Class 1 and 2) are less than 30 people which mean the sample size is too small to give reliabe satistic analysis.
Therefore, the rationality of Table_8a is doubtful.



---


# 10. Combine 3 main factors 'Gender' , 'Class' , 'Pclass', into a table sort by number of survivors.

In [195]:
#The surivial rate related to Sex, Age and Passenger class level
Table_10 = df_2[['Age Groups','Sex','Pclass','Survived']].groupby(['Sex','Pclass','Age Groups'],
                          as_index=False).sum().sort_values(by='Survived', ascending=False)
                          # Total survivors distributed in each groups
Table_10

Sex  Pclass        Age Groups  Survived
0   female       1     Adult (25-50)        52
4   female       2     Adult (25-50)        42
8   female       3     Adult (25-50)        38
12    male       1     Adult (25-50)        35
20    male       3     Adult (25-50)        29
3   female       1  Teenager (15-24)        25
11  female       3  Teenager (15-24)        20
7   female       2  Teenager (15-24)        16
2   female       1    Elder (51-100)        13
9   female       3   Children (0-14)        13
5   female       2   Children (0-14)        10
21    male       3   Children (0-14)         9
17    male       2   Children (0-14)         9
23    male       3  Teenager (15-24)         9
16    male       2     Adult (25-50)         6
14    male       1    Elder (51-100)         5
13    male       1   Children (0-14)         3
15    male       1  Teenager (15-24)         2
6   female       2    Elder (51-100)         2
1   female       1   Children (0-14)         1
18    male       2    Elder (51-100)         1
19    male       2  Teenager (15-24)         1
10  female       3    Elder (51-100)         1
22    male       3    Elder (51-100)         0

The table has same problem, small sample size can not represent true survival rate in each group set

In [0]:
df_2.to_csv('Titanic.csv',index=False)



---
## Assignment Questions



QUESTION	1:	What	categories	of	passengers	were	most	likely	to	survive	the	Titanic	disaster? à 4	points	
(the	mark	includes	2	points	for	the	answer	to	the	main	questions	and	2	points	for	the	detailed
explanation of the	logic	of	the	analysis)	

A: Female adult (age between 25 and 50) in Class '1' were most likely to survive.
From results in ***Table_1, Table_2, and Table_3b***, by looking passengers' surivial rate in each category of 'Gener', 'Age', and 'Class'.



---
QUESTION	2: What	other	attributes	did	you	use	for	the	analysis?	Explain	how	you	used	them.	Provide	a	
complete	list	of	all	attributes	used.	à 2	points


A: List of Attributes use for analysis
1. Passenger fare   *Table_3a / Table_3c*
2. Number of siblings / spouses aboard   *Table_4*
3. Number of parents / children aboard   *Table_5*
4. Port of Embarkation  *Table_6*


were all take into consideration.

Only Passenger fare has strong positive correlation with survival rate, which shows in ***Table_3a and Table_3c***



---



QUESTION	3: Did	you	engineer	any	attributes?	If	yes,	explain	the	rationale	and	how	the	new	attributes	
were	used	in	the	analysis?	If	you	have	excluded	any	attributes	from	the	analysis,	provide	an	explanation	
why	you	believe	they	have	to	be	excluded	à 2	points

A: 
**New attribute**
I create a new attribute column defined by different age stages of passenger, since grouping people of different ages stages helps to analyze the data wtih more significance.

**Exclutions:**
1. Exclude attribute "cabin" since it has too many vacancies to make reasonable analysis
2. Exclude attribute "Ticket" since the ticket numbers were just random reference numbers, and they are meaningless to survival rate



---



QUESTION	4: How	did	you	treat	missing	values?	Provide	a	detailed	explanation	in	the	comments.	à 2	
points


*   "Cabin" has more than 50% of Vacancies. Therefore, there is no necessary to replacing the Missing-value.
*   Replace all vacancies of "Age" with the average age of remaining, because age is normally distributed and there are no extremes.
*   Replace the 2 mississ entries of 'Embarked' with major Embark port 'S' beacasue port 'S' is the absolute mojority in numbers (644 out of 891)

